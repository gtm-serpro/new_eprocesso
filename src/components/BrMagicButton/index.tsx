import classNames from "classnames";
import React, { useRef } from "react";
import useCommonProperties from "../../hooks/useCommonProperties";
import { useMtProps } from "../../hooks/useMtProps";
import { useSpreadProps } from "../../hooks/useSpreadProps";
import useUniqueId from "../../hooks/useUniqueId";
import IMtProps from "../../types/IMtProps";
import BrButton from "../BrButton";

export interface BrMagicButtonProps extends React.HTMLAttributes<HTMLButtonElement>, IMtProps {
  /** Densidade do magic button.
   *
   * - large: alta
   * - medium: média
   * - small: pequena
   */
  density?: "large" | "small" | "medium";
  /** Se o botão é circular. */
  circle?: boolean;
  /** Classe de ícone FontAwesome para o botão. */
  icon?: string;
}

export interface MagicButtonRef extends HTMLButtonElement {
  element: HTMLButtonElement;
}

const BrMagicButton = React.forwardRef<MagicButtonRef, BrMagicButtonProps>(
  ({ className, id, children, density = "medium", ...props }, ref) => {
    const fid = useUniqueId(id, "magicbutton_____");
    const mtProps = useMtProps(props);
    const spreadProps = useSpreadProps(props);
    const refWrapper = useRef(null);
    const refButton = useRef<HTMLButtonElement>(null);

    useCommonProperties<MagicButtonRef>(ref, refButton, {
      get element() {
        return refButton.current;
      },
    });

    return (
      <div
        ref={refWrapper}
        className={classNames("br-magic-button", density, className, ...mtProps)}
      >
        <BrButton id={fid} ref={refButton} {...spreadProps}>
          {children}
        </BrButton>
      </div>
    );
  }
);

export default BrMagicButton;
