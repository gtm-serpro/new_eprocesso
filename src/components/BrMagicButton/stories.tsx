import { Meta, Story } from "@storybook/react";

import BrMagicButton, { BrMagicButtonProps } from ".";

export default {
  title: "Formulários/MagicButton",
  component: BrMagicButton,
} as Meta;

export const Simples: Story<BrMagicButtonProps> = (args) => (
  <BrMagicButton {...args}>Conteúdo</BrMagicButton>
);
