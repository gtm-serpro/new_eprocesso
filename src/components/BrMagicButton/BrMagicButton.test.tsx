import { render } from "@testing-library/react";

import BrMagicButton from ".";

describe("<BrMagicButton />", () => {
  it("deveria renderizar o Magic Button", () => {
    const { container } = render(<BrMagicButton>Conte√∫do</BrMagicButton>);

    expect(container.firstChild).toHaveClass("br-magic-button");
    expect(container.firstChild).toMatchSnapshot();
  });
});
