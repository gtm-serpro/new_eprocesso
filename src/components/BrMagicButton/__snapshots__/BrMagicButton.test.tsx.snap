// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BrMagicButton /> deveria renderizar o Magic Button 1`] = `
<div
  class="br-magic-button medium"
>
  <button
    class="br-button"
    id="magicbutton_____1"
    type="button"
  >
    Conte√∫do
  </button>
</div>
`;
